README
activitymanager-wheeled
=======================

Lightweight plasmoid to manage your activities effectively.
Can start, stop, switch between, clone, edit and remove activities.

This is an improvement to activitymanager plasmoid that adds the ability to switch activity by scrolling the mouse wheel over the widget.

The original plasmoid can be found here:
http://kde-apps.org/content/show.php/?content=136278

Developed by:
Abdurrahman AVCI <abdurrahmanavci AT gmail DOT com>

Improved by:
Simone Gaiarin <simgunz AT gmail DOT com>

Code under GPLv3 license (see COPYING)

Requires:
KDE SC 4.x

How to install:
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=$(kde4-config --prefix) ..
make
sudo make install


ChangeLog
0.6.2
---
* Added README with installation instructions
0.6.1
---
* Simplified wheel event management function
0.6
---
* Added ability to switch activity by scrolling mouse wheel over the widget
0.5
---
* Added GPLv2 license headers and COPYING file.
* Added tooltip to the system tray icon. Tooltip shows the
  current activity icon and name.
* Align start/stop icon to the right when activities are locked.
0.4
---
* Added ability to lock activity editing. When locked activity
  clone/edit/remove icons are hidden.
* Make sure that only one instance of rename or delete dialog
  is visible at a time. Fixes a crash.
* Use the proper activities icon.
* Added tooltips.
0.3
---
* Move action icons to the right side to be more
  consistent with the other plasmodis (e.g device notifier).
* Added ability to remove activities.
* Added ability to edit activity names.
* Added ability to clone activities.
0.2
---
* Make sure that only one extender item exist
* Sort activities by name.
0.1
---
Initial release


activitywidget.h
/***************************************************************************
* Copyright © 2010 Abdurrahman AVCI <abdurrahmanavci@gmail.com>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the
* Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
***************************************************************************/

#ifndef ACTIVITYWIDGET_H
#define ACTIVITYWIDGET_H

#include <QGraphicsWidget>
#include <QString>

namespace Plasma {
  class IconWidget;
  class LineEdit;
};

class QGraphicsGridLayout;

class ActivityWidget : public QGraphicsWidget {
  Q_OBJECT
public:
  ActivityWidget(QGraphicsItem *parent, QString id);

  QString name();
  void setName(QString name);
  void setIcon(QString icon);
  void setCurrent(bool current);
  void setState(QString state);

  void lock();
  void unlock();

public slots:
  void setCurrent();
  void toggleStatus();
  void beginRemove();
  void acceptRemove();
  void cancelRemove();
  void beginEdit();
  void acceptEdit();
  void cancelEdit();
  void beginAdd();

signals:
  void setCurrent(QString id);
  void startActivity(QString id);
  void stopActivity(QString id);
  void addActivity(QString id);
  void removeActivity(QString id);
  void renameActivity(QString id, QString name);

private:
  QGraphicsGridLayout *m_layout;
  QGraphicsWidget *m_actionsWidget;
  QGraphicsWidget *m_removeWidget;
  QGraphicsWidget *m_editWidget;
  Plasma::LineEdit *m_lineEdit;
  Plasma::IconWidget *m_label;
  Plasma::IconWidget *m_stateIcon;
  Plasma::IconWidget *m_addIcon;
  Plasma::IconWidget *m_editIcon;
  Plasma::IconWidget *m_removeIcon;
  QString m_id;
  QString m_name;
  QString m_state;
  bool m_dialogShown;
};

#endif


activitymanager.h
/***************************************************************************
* Copyright © 2010 Abdurrahman AVCI <abdurrahmanavci@gmail.com>
* Copyright © 2011 Simone Gaiarin <simgunz@gmail.com>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the
* Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
***************************************************************************/

#ifndef ACTIVITYMANAGER_H
#define ACTIVITYMANAGER_H

#include <Plasma/DataEngine>
#include <Plasma/PopupApplet>

#include <QHash>

class ActivityWidget;

class QAction;

namespace Plasma {
  class ExtenderItem;
};

class ActivityManager : public Plasma::PopupApplet {
  Q_OBJECT
public:
  ActivityManager(QObject *, const QVariantList &);

  void init();
  void initExtenderItem(Plasma::ExtenderItem *item);

public slots:
  void toolTipAboutToShow();
  void dataUpdated(QString source, Plasma::DataEngine::Data data);
  void activityAdded(QString id);
  void activityRemoved(QString id);

  void add(QString id);
  void add(QString id, QString name);
  void setCurrent(QString id);
  void stop(QString id);
  void start(QString id);
  void setName(QString id, QString name);
  void setIcon(QString id, QString name);
  void remove(QString id);

protected slots:
  virtual void wheelEvent(QGraphicsSceneWheelEvent *event);

private slots:
  void toggleLock();

private:
  void sortActivities();

  QHash<QString, ActivityWidget *> m_activities;
  QAction *toggleLockAction;
  QString m_currentName;
  QString m_currentIcon;
};

#endif


activitywidget.cpp
/***************************************************************************
* Copyright © 2010 Abdurrahman AVCI <abdurrahmanavci@gmail.com>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the
* Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
***************************************************************************/

#include "activitywidget.h"

#include <Plasma/IconWidget>
#include <Plasma/Label>
#include <Plasma/LineEdit>
#include <Plasma/PushButton>

#include <QGraphicsGridLayout>
#include <QGraphicsLinearLayout>

ActivityWidget::ActivityWidget(QGraphicsItem *parent, QString id) : QGraphicsWidget(parent), m_layout(0), m_actionsWidget(0), m_removeWidget(0), m_editWidget(0), m_label(0), m_stateIcon(0), m_addIcon(0), m_removeIcon(0), m_id(id), m_dialogShown(false) {
  m_layout = new QGraphicsGridLayout(this);
  m_layout->setContentsMargins(0, 0, 0, 0);
  setLayout(m_layout);
  // create label
  m_label = new Plasma::IconWidget(this);
  m_label->setOrientation(Qt::Horizontal);
  m_label->setPreferredSize(32, 32);
  m_label->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
  m_label->setToolTip(i18n("Click to switch to this activity."));
  m_layout->addItem(m_label, 0, 0, Qt::AlignCenter);
  // setCurrent the activity when clicked on the name
  connect(m_label, SIGNAL(clicked()), this, SLOT(setCurrent()));
  // create status icon
  m_stateIcon = new Plasma::IconWidget(this);
  m_stateIcon->setOrientation(Qt::Horizontal);
  m_stateIcon->setIcon("media-playback-start");
  m_stateIcon->setPreferredSize(22, 22);
  m_stateIcon->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
  m_stateIcon->setToolTip(i18n("Click to start this activity."));
  m_layout->addItem(m_stateIcon, 0, 1, Qt::AlignCenter);
  // toggle running state when clicked on the icon
  connect(m_stateIcon, SIGNAL(clicked()), this, SLOT(toggleStatus()));
  // create actions widget
  m_actionsWidget = new QGraphicsWidget(this);
  m_layout->addItem(m_actionsWidget, 0, 2, Qt::AlignCenter);
  // create a layout
  QGraphicsGridLayout *layout = new QGraphicsGridLayout(m_actionsWidget);
  layout->setContentsMargins(0, 0, 0, 0);
  m_actionsWidget->setLayout(layout);
  // create add icon
  m_addIcon = new Plasma::IconWidget(this);
  m_addIcon->setOrientation(Qt::Horizontal);
  m_addIcon->setSvg("widgets/action-overlays", "add-normal");
  m_addIcon->setPreferredSize(22, 22);
  m_addIcon->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
  m_addIcon->setToolTip(i18n("Click to clone this activity."));
  layout->addItem(m_addIcon, 0, 2, Qt::AlignCenter);
  // begin adding
  connect(m_addIcon, SIGNAL(clicked()), this, SLOT(beginAdd()));
  // create edit icon
  m_editIcon = new Plasma::IconWidget(this);
  m_editIcon->setOrientation(Qt::Horizontal);
  m_editIcon->setSvg("widgets/configuration-icons", "configure");
  m_editIcon->setPreferredSize(16, 16);
  m_editIcon->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
  m_editIcon->setToolTip(i18n("Click to edit this activity."));
  layout->addItem(m_editIcon, 0, 3, Qt::AlignCenter);
  // begin editing
  connect(m_editIcon, SIGNAL(clicked()), this, SLOT(beginEdit()));
  // create remove icon
  m_removeIcon = new Plasma::IconWidget(this);
  m_removeIcon->setOrientation(Qt::Horizontal);
  m_removeIcon->setSvg("widgets/configuration-icons", "close");
  m_removeIcon->setPreferredSize(16, 16);
  m_removeIcon->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
  m_removeIcon->setToolTip(i18n("Click to remove this activity."));
  layout->addItem(m_removeIcon, 0, 4, Qt::AlignCenter);
  // ask for confirmation
  connect(m_removeIcon, SIGNAL(clicked()), this, SLOT(beginRemove()));
}

QString ActivityWidget::name() {
  return m_name;
}

void ActivityWidget::setName(QString name) {
  m_name = name;
  // update label
  m_label->setText(m_name);
}

void ActivityWidget::setIcon(QString icon) {
  // update label
  if (icon != "")
    m_label->setIcon(icon);
  else
    m_label->setIcon("plasma");
}

void ActivityWidget::setCurrent(bool current) {
  // update label
  QFont font = m_label->font();
  if (current)
    font.setWeight(QFont::Bold);
  else
    font.setWeight(QFont::Normal);
  m_label->setFont(font);
}

void ActivityWidget::setState(QString state) {
  m_state = state;
  // update icon
  if (m_state == "Stopped") {
    m_stateIcon->setIcon("media-playback-start");
    m_stateIcon->setToolTip(i18n("Click to start this activity."));
  } else {
    m_stateIcon->setIcon("media-playback-stop");
    m_stateIcon->setToolTip(i18n("Click to stop this activity."));
  }
}

void ActivityWidget::lock() {
  m_layout->removeAt(2);
  m_actionsWidget->setVisible(false);
}

void ActivityWidget::unlock() {
  m_actionsWidget->setVisible(true);
  m_layout->addItem(m_actionsWidget, 0, 2, Qt::AlignCenter);
}

void ActivityWidget::setCurrent() {
  emit setCurrent(m_id);
}

void ActivityWidget::toggleStatus() {
  if (m_state == "Stopped")
    emit startActivity(m_id);
  else
    emit stopActivity(m_id);
}

void ActivityWidget::beginRemove() {
  if (m_dialogShown)
    return;
  // turn on the removing flag
  m_dialogShown = true;
  // create confirmation widget
  m_removeWidget = new QGraphicsWidget();
  // create a horizontal layout
  QGraphicsLinearLayout *layout = new QGraphicsLinearLayout(m_removeWidget);
  layout->setOrientation(Qt::Horizontal);
  // set the layout
  m_removeWidget->setLayout(layout);
  // create label
  Plasma::Label *label = new Plasma::Label(m_removeWidget);
  label->setText(i18n("Do you really want to remove this activity?"));
  layout->addItem(label);
  // create yes button
  Plasma::PushButton *yesButton = new Plasma::PushButton(m_removeWidget);
  yesButton->setText(i18n("Yes"));
  layout->addItem(yesButton);
  connect(yesButton, SIGNAL(clicked()), this, SLOT(acceptRemove()));
  // create no button
  Plasma::PushButton *noButton = new Plasma::PushButton(m_removeWidget);
  noButton->setText(i18n("No"));
  layout->addItem(noButton);
  connect(noButton, SIGNAL(clicked()), this, SLOT(cancelRemove()));
  // show the confirm widget
  m_layout->addItem(m_removeWidget, 1, 0, 1, 5);
}

void ActivityWidget::acceptRemove() {
  // delete the confirm widget
  m_removeWidget->deleteLater();
  // emit delete signal
  emit removeActivity(m_id);
  // turn off the removing flag
  m_dialogShown = false;
}

void ActivityWidget::cancelRemove() {
  // delete the confirm widget
  m_removeWidget->deleteLater();
  // turn off the removing flag
  m_dialogShown = false;
}

void ActivityWidget::beginEdit() {
  if (m_dialogShown)
    return;
  // turn on the editing flag
  m_dialogShown = true;
  // create confirmation widget
  m_editWidget = new QGraphicsWidget();
  // create a horizontal layout
  QGraphicsLinearLayout *layout = new QGraphicsLinearLayout(m_editWidget);
  layout->setOrientation(Qt::Horizontal);
  // set the layout
  m_editWidget->setLayout(layout);
  // create label
  Plasma::Label *label = new Plasma::Label(m_editWidget);
  label->setText(i18n("Name:"));
  layout->addItem(label);
  // create the line edit
  m_lineEdit = new Plasma::LineEdit(m_editWidget);
  m_lineEdit->setText(m_name);
  layout->addItem(m_lineEdit);
  // create yes button
  Plasma::PushButton *yesButton = new Plasma::PushButton(m_editWidget);
  yesButton->setText(i18n("OK"));
  layout->addItem(yesButton);
  connect(yesButton, SIGNAL(clicked()), this, SLOT(acceptEdit()));
  // create no button
  Plasma::PushButton *noButton = new Plasma::PushButton(m_editWidget);
  noButton->setText(i18n("Cancel"));
  layout->addItem(noButton);
  connect(noButton, SIGNAL(clicked()), this, SLOT(cancelEdit()));
  // show the confirm widget
  m_layout->addItem(m_editWidget, 1, 0, 1, 5);
  // put the focus into the line edit
  m_lineEdit->setFocus(Qt::OtherFocusReason);
}

void ActivityWidget::acceptEdit() {
  // save the name
  QString name = m_lineEdit->text();
  // delete the edit widget
  m_editWidget->deleteLater();
  // emit rename signal
  emit renameActivity(m_id, name);
  // turn of the editing flag
  m_dialogShown = false;
}

void ActivityWidget::cancelEdit() {
  // delete the edit widget
  m_editWidget->deleteLater();
  // turn off the editing flag
  m_dialogShown = false;
}

void ActivityWidget::beginAdd() {
  // emit add signal
  emit addActivity(m_id);
}


activitymanager.cpp
/***************************************************************************
* Copyright © 2010 Abdurrahman AVCI <abdurrahmanavci@gmail.com>
* Copyright © 2011 Simone Gaiarin <simgunz@gmail.com>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the
* Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
***************************************************************************/

#include "activitymanager.h"

#include "activitywidget.h"

#include <Plasma/Extender>
#include <Plasma/ExtenderItem>
#include <Plasma/Service>
#include <Plasma/ServiceJob>
#include <Plasma/ToolTipContent>
#include <Plasma/ToolTipManager>

#include <QAction>
#include <QGraphicsLinearLayout>
#include <QString>
#include <QGraphicsSceneWheelEvent>

ActivityManager::ActivityManager(QObject *parent, const QVariantList &args): Plasma::PopupApplet(parent, args), toggleLockAction(0), m_currentName(""), m_currentIcon("") {
  setPopupIcon("preferences-activities");
  setAspectRatioMode(Plasma::IgnoreAspectRatio);
}

void ActivityManager::init() {
  Plasma::ToolTipManager::self()->registerWidget(this);
  extender()->setEmptyExtenderMessage(i18n("No Activities..."));
  // don't grow too much height
  extender()->setMaximumHeight(300);
  if (extender()->item("Activities") == 0) {
    // create the item
    Plasma::ExtenderItem *item = new Plasma::ExtenderItem(extender());
    // initialize the item
    initExtenderItem(item);
    // set item name and title
    item->setName("Activities");
    item->setTitle("Activities");
  }
  // connect data sources
  Plasma::DataEngine *engine = dataEngine("org.kde.activities");
  foreach (const QString source, engine->sources())
    activityAdded(source);
  // activity addition and removal
  connect(engine, SIGNAL(sourceAdded(QString)), this, SLOT(activityAdded(QString)));
  connect(engine, SIGNAL(sourceRemoved(QString)), this, SLOT(activityRemoved(QString)));
}

void ActivityManager::initExtenderItem(Plasma::ExtenderItem *item) {
  // create the widget
  QGraphicsWidget *widget = new QGraphicsWidget(this);
  // TODO: use the size of the longest activity name
  widget->setPreferredWidth(350);
  // create the layout
  QGraphicsLinearLayout *layout = new QGraphicsLinearLayout(widget);
  layout->setOrientation(Qt::Vertical);
  widget->setLayout(layout);
  // set up the widget
  item->setWidget(widget);
  // create a lock/unlock action
  toggleLockAction = new QAction(item);
  toggleLockAction->setIcon(KIcon("object-locked"));
  toggleLockAction->setEnabled(true);
  toggleLockAction->setVisible(true);
  toggleLockAction->setToolTip(i18n("Activities are unlocked. Click to lock."));
  item->addAction("toggleLock", toggleLockAction);
  connect(toggleLockAction, SIGNAL(triggered()), this, SLOT(toggleLock()));
}

void ActivityManager::toolTipAboutToShow() {
  Plasma::ToolTipContent toolTip;

  toolTip.setMainText(i18n("Current Activity: %1").arg(m_currentName));
  if (!m_currentIcon.isEmpty())
    toolTip.setImage(KIcon(m_currentIcon));

  Plasma::ToolTipManager::self()->setContent(this, toolTip);
}

void ActivityManager::dataUpdated(QString source, Plasma::DataEngine::Data data) {
  if (!m_activities.contains(source))
    return;
  ActivityWidget *activity = m_activities[source];
  // update activity info
  activity->setName(data["Name"].toString());
  activity->setState(data["State"].toString());
  activity->setIcon(data["Icon"].toString());
  activity->setCurrent(data["Current"].toBool());
  // update current activity name and icon
  if (data["Current"].toBool()) {
    m_currentName = data["Name"].toString();
    m_currentIcon = data["Icon"].toString();
  }
  // sort activities
  sortActivities();
}

void ActivityManager::activityAdded(QString id) {
  // skip the Status source
  if (id == "Status")
    return;
  // create a new activity object
  ActivityWidget *activity = new ActivityWidget(extender()->item("Activities"), id);
  // add activity to the list
  m_activities.insert(id, activity);
  // connect activity update signal
  dataEngine("org.kde.activities")->connectSource(id, this);
  // connect activity start/stop signals
  connect(activity, SIGNAL(setCurrent(QString)), this, SLOT(setCurrent(QString)));
  connect(activity, SIGNAL(startActivity(QString)), this, SLOT(start(QString)));
  connect(activity, SIGNAL(stopActivity(QString)), this, SLOT(stop(QString)));
  connect(activity, SIGNAL(addActivity(QString)), this, SLOT(add(QString)));
  connect(activity, SIGNAL(removeActivity(QString)), this, SLOT(remove(QString)));
  connect(activity, SIGNAL(renameActivity(QString,QString)), this, SLOT(setName(QString,QString)));
}

void ActivityManager::activityRemoved(QString id) {
  if (!m_activities.contains(id))
    return;
  // delete the activity
  delete m_activities.take(id);
}

void ActivityManager::add(QString id) {
  add(id, "New Activity");
}

void ActivityManager::add(QString id, QString name) {
  Plasma::Service *service = dataEngine("org.kde.activities")->serviceForSource(id);
  KConfigGroup op = service->operationDescription("add");
  op.writeEntry("Name", name);
  Plasma::ServiceJob *job = service->startOperationCall(op);
  connect(job, SIGNAL(finished(KJob*)), service, SLOT(deleteLater()));
}

void ActivityManager::setCurrent(QString id) {
  Plasma::Service *service = dataEngine("org.kde.activities")->serviceForSource(id);
  Plasma::ServiceJob *job = service->startOperationCall(service->operationDescription("setCurrent"));
  connect(job, SIGNAL(finished(KJob*)), service, SLOT(deleteLater()));
}

void ActivityManager::stop(QString id) {
  // TODO: when activity is stopped, take a screenshot and use that icon
  Plasma::Service *service = dataEngine("org.kde.activities")->serviceForSource(id);
  Plasma::ServiceJob *job = service->startOperationCall(service->operationDescription("stop"));
  connect(job, SIGNAL(finished(KJob*)), service, SLOT(deleteLater()));
}

void ActivityManager::start(QString id) {
  Plasma::Service *service = dataEngine("org.kde.activities")->serviceForSource(id);
  Plasma::ServiceJob *job = service->startOperationCall(service->operationDescription("start"));
  connect(job, SIGNAL(finished(KJob*)), service, SLOT(deleteLater()));
}

void ActivityManager::setName(QString id, QString name) {
  Plasma::Service *service = dataEngine("org.kde.activities")->serviceForSource(id);
  KConfigGroup op = service->operationDescription("setName");
  op.writeEntry("Name", name);
  Plasma::ServiceJob *job = service->startOperationCall(op);
  connect(job, SIGNAL(finished(KJob*)), service, SLOT(deleteLater()));
}

void ActivityManager::setIcon(QString id, QString name) {
  Plasma::Service *service = dataEngine("org.kde.activities")->serviceForSource(id);
  KConfigGroup op = service->operationDescription("setIcon");
  op.writeEntry("Icon", name);
  Plasma::ServiceJob *job = service->startOperationCall(op);
  connect(job, SIGNAL(finished(KJob*)), service, SLOT(deleteLater()));
}

void ActivityManager::remove(QString id) {
  Plasma::Service *service = dataEngine("org.kde.activities")->serviceForSource(id);
  KConfigGroup op = service->operationDescription("remove");
  op.writeEntry("Id", id);
  Plasma::ServiceJob *job = service->startOperationCall(op);
  connect(job, SIGNAL(finished(KJob*)), service, SLOT(deleteLater()));
}

void ActivityManager::toggleLock() {
  if (toggleLockAction->toolTip() == i18n("Activities are locked. Click to unlock.")) {
    toggleLockAction->setIcon(KIcon("object-unlocked"));
    toggleLockAction->setToolTip(i18n("Activities are unlocked. Click to lock."));
    // unlock activities
    foreach (ActivityWidget *activity, m_activities)
      activity->unlock();
  } else {
    toggleLockAction->setIcon(KIcon("object-locked"));
    toggleLockAction->setToolTip(i18n("Activities are locked. Click to unlock."));
    // lock activities
    foreach (ActivityWidget *activity, m_activities)
      activity->lock();
  }
}

void ActivityManager::sortActivities() {
  QGraphicsLinearLayout *layout = static_cast<QGraphicsLinearLayout *>(static_cast<QGraphicsWidget *>(extender()->item("Activities")->widget())->layout());
  // remove all activities from the layout
  while (layout->count())
    layout->removeAt(0);
  // insert all activities to a map
  QMap<QString, ActivityWidget *> activities;
  foreach (ActivityWidget *activity, m_activities.values())
    activities.insertMulti(activity->name(), activity);
  // add activities into the layout at ascending order of their names
  foreach (ActivityWidget *activity, activities.values())
    layout->addItem(activity);
}

void ActivityManager::wheelEvent(QGraphicsSceneWheelEvent *e)
{
    Plasma::DataEngine *engine = dataEngine("org.kde.activities");
    Plasma::DataEngine::Data data = engine->query("Status");
    QStringList activities = data["Running"].toStringList();
    QString currentActivity = data["Current"].toString();

    int nextIndex = activities.indexOf(currentActivity);

    if(e->delta() < 0)
        nextIndex = (nextIndex + 1) % activities.size();
    else
    {
        if(nextIndex - 1 < 0)
            nextIndex = activities.size() - 1;
        else
            nextIndex = nextIndex - 1;
    }

    Plasma::Service *service = engine->serviceForSource(activities[nextIndex]);
    service->startOperationCall(service->operationDescription("setCurrent"));
}

K_EXPORT_PLASMA_APPLET(activitymanager, ActivityManager)
